Add biotech support

0x00002a opened the pull request:
Hi

From your comment in #6 I gather that you arn't planning on adding support for biotech, so I decided to do it instead.

## Patched buildings

- Band node (unbound nodes consume no power)
- Gene assembler and processors (processors will consume power only when assembler is in use), gene banks still always consume power since they are for storage
- Growth vat 
- Mech charger
- Mech gestator (fixes #6 as well)
- Subcore scanners (rip + soft)
- Waste atomizer

Note I haven't tested this without biotech, and I haven't tested it that thoughly beyond checking stuff \*seems\* to work at a surface level. I'm going to be playing with these changes anyway over the next few days so I'll report back and fix anything I find thats broken.

NathanIkola commented on Source/LightsOutSln/LightsOut/Patches/ModCompatibility/Biotech/PatchMechCharger.cs:
```diff
@@ -0,0 +1,67 @@
+
+        private static void AfterSpawn(Building __instance)
+        {
+            if (IsAttachedToMech == null)
```
Realistically I could go either way on this since it's probably a micro-optimization, but having this check here guarantees that we do the ref equals check on this property for every building spawn in the game. If you move this if statement inside of the __instance is Building_MechCharger condition then we will only pay for it for the mech chargers.

Better yet, you could start this method out with `if (!(__instance is Building_MechCharger ch)) { return; }` so that everything else exits as early as possible.

NathanIkola commented on Source/LightsOutSln/LightsOut/Patches/ModCompatibility/Biotech/PatchSubscoreScanner.cs:
```diff
@@ -0,0 +1,27 @@
+
+namespace LightsOut.Patches.ModCompatibility.Biotech
+{
+    public class PatchSubscoreScanner : ICompatibilityPatchComponent<PatchSubscoreScanner>
```
There's a typo in your type name

NathanIkola commented on Source/LightsOutSln/LightsOut/Patches/ModCompatibility/Biotech/PatchWasteAtomiser.cs:
```diff
@@ -0,0 +1,43 @@
+            {
+                new PatchInfo
+                {
+                    method = GetMethod<CompAtomizer>(nameof(CompAtomizer.CompTick)),
```
Avoid patching tickwise methods at all cost, they are more expensive than they seem.

You are likely better off patching `CompAtomizer`'s `Notify_MaterialsAdded`, `EjectContents`, and `DoAtomize` methods to cover your functionality.

NathanIkola commented on Source/LightsOutSln/LightsOut/Patches/ModCompatibility/Biotech/PatchGeneAssembly.cs:
```diff
@@ -0,0 +1,70 @@
+    public class PatchGeneAssembly : ICompatibilityPatchComponent<PatchGeneAssembly>
+    {
+        public override string ComponentName => "Patch for gene assembly related things";
+        private const string PROCESSOR_ID = "GeneProcessor";
```
The other constants in this mod use pascal case, and this name isn't very descriptive.

I'd rename it something like `GeneProcessorDefName`.

NathanIkola reviewed (changes requested):
In addition to the above comments, I have some various notes that are wider-spread.

You should add the XML documentation above your classes, methods, and members. This is consumed by the documentation generator I use for the mod, so not having those means we're lacking some documentation that we could otherwise have very easily.

All of your patch components are declared with `: ICompatibilityPatchComponent<YourPatchName>` which is not how that should be used. The template type specified is the type you're supposed to be actually patching. It doesn't make a functional difference with your patches, but it's not how they're supposed to be used. My guess is you should be using these instead:

PatchBandNode -> CompBandNode
PatchGeneAssembly -> Building_GeneAssembler
PatchGeneExtractor -> Building_GeneExtractor
PatchGrowthVat -> Building_GrowthVat
PatchMechCharger -> Building_MechCharger
PatchMechGestator -> Building_MechGestator
PatchSubcoreScanner -> Building_SubcoreScanner
PatchWasteAtomiser -> CompAtomizer

0x00002a commented:
Thanks for the comments, I'll see about implementing them soon:tm:

NathanIkola commented:
Yikes, on rereading my comments they could come across as pretty dry/aggressive-sounding (it's probably a commentary on how I sound when I do code review at work, too, but there's no way I'm doing any personal reflection on a Monday morning). And I never even thanked you for your contribution. Sorry about that, and thanks for the help on this!

The code in here looks really clean and well-written, and I very much appreciate how you structured everything. Overall I'm excited to get Biotech support into the mod, and the community is going to appreciate it even more! I'm sure several of my comments above will come down to me simply not understanding Biotech things. With that in mind, feel free to push back on any of the feedback I give in general; I'm totally happy to have discussions about things, and I don't want to come across as a "my way or the highway" kind of person. There are countless places in my code that could use touch-ups (some of which your changes highlight very well), and probably many times that more bugs lurking about; a fresh perspective and discussion will only benefit the mod's quality.

Thanks again, and happy holidays!

0x00002a commented:
Apologies for the long response time, I've had exams and holiday time :p.

I've addressed all your comments I think. I've added a new `Enterable` helper to handle the things that should be enabled on spawn if they have an ocupant but arn't tables. I've also made `Tables` use a `HashSet` lookup for table types so we don't have to patch `IsTable` anymore.

Theres also a fix for the mech gestator going idle when the gestation cycle was advanced (issue with it being identified as a table so the job "finishing" turning it off despite it still gestating the mech)

0x00002a commented:
Features currently missing (that I totally didn't just notice a minute ago):

- deathrest casket

NathanIkola commented:
FYI, this introduces a bug in the table logic; it now only detects tables that are explicitly of type Building_WorkTable or Building_ResearchBench, so tables (like the electric stove) no longer get activated.

Fixed in commit [09212f5](https://github.com/NathanIkola/LightsOut/commit/09212f55e9b8aa94aab54d98ca574063129552d9).
